### PR TITLE
Add `-Wl,-rpath-link` for secondary dependencies

### PR DESCRIPTION
**Pull request by SoapGentoo**
@nirbheek Revert my fix commit and the test case will fail on current `master`

### CODE REVIEW

**Comment by jpakkane** [comment]
The LD manual recommends against using `rpath-link` but I don't really see any other way of doing this. I'd appreciate a second set of eyes, maybe @nirbheek or @dcbaker?

**Comment by textshell** [comment]
I haven't yet looked at the implementation but i've added a note to #3111 some time ago about this:

> Dynamic linking
At least gnu binutils' ld insists that dependencies of shared libraries (DT_NEEDED) need to be available and findable at link time. Practically this means all locations (containing directory) of secondary dependencies need to be either specified either as -rpath or -rpath-link
> * Currently not handled at all

So yes, we need either -rpath or -rpath-link and i believe -rpath-link is not harmful even if for other reasons -rpath is needed for an path as well.

**Comment by textshell** [comment]
We need to make sure we have deduplication support for this. Otherwise this will be a performance regression.

File: mesonbuild/dependencies/base.py [hunk 1]
```diff
@@ -753,15 +755,31 @@ def _search_libs(self, out, out_raw):
...
         # Track -lfoo libraries to avoid duplicate work
         libs_found = OrderedSet()
         # Track not-found libraries to know whether to add library paths
         libs_notfound = []
         libtype = LibType.STATIC if self.static else LibType.PREFER_SHARED
-        # Generate link arguments for this library
+        # Generate link arguments for this library, by
+        # first appending secondary link arguments for ld
         link_args = []
+        if self.clib_compiler and get_compiler_is_linuxlike(self.clib_compiler):
```
Comment (dcbaker) [inline mesonbuild/dependencies/base.py]: lets not use `get_compiler_is_linuxlike` please. I'm trying to delete that. Something like:
```python
if self.clib_compiler and self.clib.compiler.linker and isinstance(self.clib_compiler.linker, linkers.GnuLikeDynamicLinkerMixin):
```

Comment (SoapGentoo) [reply mesonbuild/dependencies/base.py]: `linkers.GnuLikeDynamicLinkerMixin` exists nowhere in the Meson codebase

Comment (dcbaker) [reply mesonbuild/dependencies/base.py]: you need to rebase, it's in mesonbuild/linkers.py

**Comment by lantw44** [comment]
This breaks secondary dependencies finding and causes a lot of undefined reference errors under JHBuild. Please see https://github.com/mesonbuild/meson/issues/6027.

**Pull request merged**